### DISCUSSION TITLE
Throw own exception in catch block of Zod objects

### DISCUSSION THREAD

**Discussion by zackarydev** [colinhacks/zod#2993] [Q&A] [answered]
Instead of returning a value in a catch block, I'd rather throw my own exception.

Right now doing this would result in the `z.ZodCatch` type being assigned as a wrapper.

Ex:

```ts
export const IdField = z
    .string()
    .uuid()
    .readonly()
    .catch((ctx): string => {
        throw new InvalidIdError('v4', ctx.input);
    });
```

Although this works by itself, when using `z.merge` later on, this breaks in the input field of the context of another object.

```ts
const MyObject = z
    .object({
        id: IdField,
        name: z.string(),
    })
    .catch((ctx) => {
         throw new InvalidObjectError(ctx.input.id); <-- id is unknown
    })
```

Is there anyway we can do this in Zod?

IMO this is because ZodCatch doesn't extend ZodObject so we can't chain them together :(

**Comment by zackarydev** [comment]
All I'd want to do though is `MyObject.parse()` and customize the parsing error

**Comment by JacobWeisenburger** [comment]
Is this what you are looking for?

```ts
function parseOrThrow ( schema: z.Schema, data: unknown, error: Error ) {
    const parsed = schema.safeParse( data )
    if ( parsed.success ) return parsed.data
    throw error
}

class MyCustomError extends Error {
    /* do whatever you what */
}

try {
    console.log(
        parseOrThrow(
            z.string(),
            'foo',
            new MyCustomError( 'My Custom Error' )
        )
    ) // 'foo'
} catch ( error ) {
    // no error
}

try {
    parseOrThrow(
        z.string(),
        42,
        new MyCustomError( 'My Custom Error' )
    ) // throws
} catch ( error ) {
    console.error( ( error as Error ).message ) // 'My Custom Error'
}
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by zackarydev** [reply]
I mean you can do these `parseOrThrow` but when you have a Zod object 15 layers deep we don't want to create a function for each key/value of the 15 nested objects...

We would have expected that attaching the `.catch` to the field's Zod object would have been enough to catch the exception and allow us to throw our own.

**Reply by zackarydev** [reply]
These examples above work well for a single field, but as soon as you start composing bigger objects they become .... unless.

**Comment by JacobWeisenburger** [comment, marked as answer]
Is this what you are looking for?

```ts
class MyCustomError extends Error {
    /* do whatever you what */
}

const schema = z.object( {
    foo: z.string( { errorMap: () => ( { message: 'Custom Foo Error' } ) } ),
    bar: z.number( { errorMap: () => ( { message: 'Custom Bar Error' } ) } ),
} )

const result = schema.safeParse( {} )
const customErrors = !result.success
    ? result.error.issues.map( ( { message } ) => new MyCustomError( message ) )
    : []
console.log( customErrors.map( ( { message } ) => message ) )
// [ "Custom Foo Error", "Custom Bar Error" ]
```

If you found my answer satisfactory, please consider supporting me. Even a small amount is greatly appreciated. Thanks friend! 🙏
https://github.com/sponsors/JacobWeisenburger

**Reply by zackarydev** [reply]
I think this is definitely closer. But instead of a string the ability to return any object/a subclass of Error
